[Feature] Updates audio devices list when connected to bluetooth/headphones

cheukchuen opened the pull request:
## Purpose
<!-- Describe the intention of the changes being proposed. What problem does it solve or functionality does it add? -->
* Replaces "iPhone/iPad" with "Headphones" when connected to headphones
* Replaces "iPhone/iPad" with "Bluetooth (icon name)" when connected to bluetooth
* Replaces "iPhone/iPad" with "\<bluetooth device name\> (audio list)" when connected to bluetooth

## Does this introduce a breaking change?
<!-- Mark one with an "x". -->
```
[ ] Yes
[x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## Other Information
<!-- Add any other helpful information that may be needed here. -->
![IMG_0017](https://user-images.githubusercontent.com/11702891/156222687-1c26fa9f-3363-4844-a480-87e374dd0ff7.PNG)
![IMG_0018](https://user-images.githubusercontent.com/11702891/156222685-a00d6679-ebd3-47af-b8b4-b847600559f1.PNG)
![IMG_0019](https://user-images.githubusercontent.com/11702891/156222683-fba51116-5535-4672-997f-aacf6edc60b2.PNG)
![IMG_0020](https://user-images.githubusercontent.com/11702891/156222680-004b4acc-3116-4716-8ce2-1f270878c2d3.PNG)

cheukchuen committed:
Updates audio devices list when connected to bluetooth/headphones

cheukchuen committed:
Refactor drawer viewControllers for audio devices list and participants list

cheukchuen committed:
Update unit tests for AudioDeviceListViewModel

cheukchuen committed:
Merge branch 'develop/beta' into feature/audio-devices-detection

cheukchuen committed:
Fix merge conflict in updated names

jimchou-dev commented:
Looks like the Text label is a bit cut off
<img width="288" alt="image" src="https://user-images.githubusercontent.com/77021369/156461006-82ca82da-aa4b-485d-9266-cd9141771e70.png">

jimchou-dev commented on AzureCommunicationUI/AzureCommunicationUI/Presentation/SwiftUI/ViewComponents/Drawer/AudioDevicesListViewModel.swift:
```diff
@@ -0,0 +1,80 @@
+        default:
+            audioDeviceType = .receiver
+        }
+        audioDeviceOptions.append(getAudioDeviceOption(for: audioDeviceType))
+        audioDeviceOptions.append(getAudioDeviceOption(for: .speaker))
```
Would it be possible for the audio device to have speaker/wired headphone/bluetooth headphone so have more than 2 table cell?

cheukchuen replied:
The iOS audio override API provides us with 2 options: `speaker` and `none`, where `none` means default system audio that's currently being used, hence we provide 2 options for the audio device selection.

In reference, Teams also provide 2 options as follows (ignore the `Audio off` since we do not support it):
![IMG_0025](https://user-images.githubusercontent.com/11702891/156681366-8f55a56e-f248-4b2d-82f4-b90b4d14cb47.PNG)
![IMG_0026](https://user-images.githubusercontent.com/11702891/156681365-c22c5931-19b9-43b2-89a6-96bbc82b8bff.PNG)

jimchou-dev commented on AzureCommunicationUI/AzureCommunicationUI/Presentation/Manager/AudioSessionManager.swift:
```diff
@@ -66,8 +70,21 @@ class AppAudioSessionManager: AudioSessionManager {
-            return .speaker
+        if let output = audioSession.currentRoute.outputs.first {
+            switch output.portType {
+            case .bluetoothA2DP:
+                return .bluetooth
+            case .bluetoothLE:
+                return .bluetooth
+            case .bluetoothHFP:
+                return .bluetooth
```
`case .bluetoothA2DP, .bluetoothLE, .bluetoothHFP:`

kevinyulu commented on AzureCommunicationUI/AzureCommunicationUI/Presentation/FluentUI/Wrapper/DrawerContainerViewController.swift:
```diff
@@ -65,11 +79,42 @@ class DrawerContainerViewController<T>: UIViewController, DrawerControllerDelega
+            let tableCellsDividerOffsetHeight = CGFloat(self.items.count * 3)
+            drawerHeight += tableCellsDividerOffsetHeight + self.drawerResizeBarHeight
+        } else {
+            drawerHeight = CGFloat(self.items.count) * 48.5
```
for iPad is 48.5 minimum height for drawer to show all the content? curious on how the number is calculated

cheukchuen replied:
It was for a fix pre- public release. Previously we used `participantsListTableView.contentSize.height` to dynamically obtain the cell height, but found out that it doesn't stay constant and flips between (if I recall correctly) 44 and 48.5 when rendering the drawer. So instead we specifically set the base row height to 44 for iPhone and 48.5 for iPad after some trail and error by observing `participantsListTableView.contentSize.height` during runtime.

kevinyulu commented on AzureCommunicationUI/AzureCommunicationUI/Presentation/FluentUI/Wrapper/CompositeAudioDevicesListCell.swift:
```diff
@@ -0,0 +1,35 @@
+            ? StyleProvider.color.popoverColor
+            : StyleProvider.color.drawerColor
+
+        setTitleLabelTextColor(color: UIColor.compositeColor(CompositeColor.onSurface))
```
should the selectionStyle property also be `.none`? I notice when the whole cell is highlighted, with pressed. Was this is a recent UX update?

cheukchuen replied:
Thanks for pointing that out. Made the necessary change to reflect the previous UX.

Also specified `selectionStyle = .none` for participants list since we may want to add interaction to the list in the future.

kevinyulu commented:
Tested the app with a few test scenarios, bluetooth and headphone works in the call, with a few observations:
Test scenario (connected to headphone or bluetooth headset before the call):
* When switching audio type from `Speaker` to `Headphone`, the title of the first cell item in audio device list was `iPhone`, and then becomes `Headphone` quickly. When switching back to `Speaker`, the cell title changed back to `iPhone`. Not sure if this name should stay at `Headphone` in this case. 
The same observed for the case of switching back and forth between `Bluetooth headset` and `Speaker`. 

Test scenario (connected to bluetooth headset before the call):
* When switching audio type from `Bluetooth Headset` to `Headphone`, the icon of the first cell item in audio device list was updated to the new "speaker with bluetooth" icon. Should this new icon be updated in the audio switch icon the control bar button? maybe that's a UX question.

cheukchuen committed:
Address PR comments

- expand long icon button text
- handle scenarios when switching to speaker while headphones and/or bluetooth headset are connected
- other misc improvements

cheukchuen committed:
Merge branch 'develop/beta' into feature/audio-devices-detection

cheukchuen committed:
Do not style cell selection on pressing

cheukchuen commented on AzureCommunicationUI/AzureCommunicationUI/Presentation/FluentUI/Wrapper/CompositeAudioDevicesList.swift:
```diff
@@ -6,30 +6,34 @@
-        let controller = PopupMenuViewController(items: viewModel.audioDeviceList,
-                                                 sourceView: sourceView)
+    func makeUIViewController(context: Context) -> DrawerContainerViewController<AudioDevicesListCellViewModel> {
+        let controller = AudioDevicesListViewController(items: getAudioDevicesList(),
```
So previously we utilized `PopupMenuController` to display the audio device selection, though the API they provide only limits us to adding new items to the drawer. Limitation being when we want to update or remove items from the list since there's no direct methods to do so, even for accessing the items list since those are private fields.

There may be workaround to manipulating the drawer items by indirectly accessing the UITable from `PopupMenuController`, but it's convoluted and also need to manually update the UI when the height changes. Using the tableViewController used for participants list removes all these issues and also maintainable since our drawer lists now use the same tableViewController.

cheukchuen commented:
> Tested the app with a few test scenarios, bluetooth and headphone works in the call, with a few observations: Test scenario (connected to headphone or bluetooth headset before the call):
> 
> * When switching audio type from `Speaker` to `Headphone`, the title of the first cell item in audio device list was `iPhone`, and then becomes `Headphone` quickly. When switching back to `Speaker`, the cell title changed back to `iPhone`. Not sure if this name should stay at `Headphone` in this case.
>   The same observed for the case of switching back and forth between `Bluetooth headset` and `Speaker`.
> 
> Test scenario (connected to bluetooth headset before the call):
> 
> * When switching audio type from `Bluetooth Headset` to `Headphone`, the icon of the first cell item in audio device list was updated to the new "speaker with bluetooth" icon. Should this new icon be updated in the audio switch icon the control bar button? maybe that's a UX question.

@kevinyulu
1. `Headphones`/`Bluetooth` should stay the same when switching to `Speaker`. Made some code changes to account for this behavior, please try it out and let me know if there's anything I missed.
2. The speaker icon on the control bar should reflect on the selected audio device, both before the call in setup and during the call. So icon should be bluetooth speaker when connected to bluetooth headset, and regular speaker (same as receiver) when connected to headphones. Please let me know if there's a bug in the control bar as well as the selection list and I'll proceed to fix them.

cheukchuen commented:
> Looks like the Text label is a bit cut off <img alt="image" width="288" src="https://user-images.githubusercontent.com/77021369/156461006-82ca82da-aa4b-485d-9266-cd9141771e70.png">

@jimchou-dev I've expanded the width of the IconButton to 80 to cater specifically for `Headphones`. In the future when we localize those strings, it may be even longer than the allotted space so we can't avoid ellipsis indefinitely so something to consider.

AmyL219 commented on AzureCommunicationUI/AzureCommunicationUI/Presentation/Manager/AudioSessionManager.swift:
```diff
@@ -77,10 +90,10 @@ class AppAudioSessionManager: AudioSessionManager {
 
         let audioPort: AVAudioSession.PortOverride
```
Why don't we handle .headphone, bluetooch, speaker here?

cheukchuen replied:
`func overrideOutputAudioPort(_ portOverride: AVAudioSession.PortOverride)` accepts 2 options, `.speaker` and `.none`, where `.none` represents the audio output that the system default is currently in, hence we provide iPhone/Headphones/Bluetooth as one of the selections in our audio devices list based on what has been connected.

Not sure if that answers your question though, let me know if we need to follow up on this.

AmyL219 replied:
Okey, typically for  iPhone/Headphones/Bluetooth we would set them as `.none`? Possibly we can indicate that in the switch
```
 switch selectedAudioDevice {
        case .speaker:
            audioPort = .speaker
        case .bluetooth,
                .receiver, 
                headphones,
                etc:
            audioPort = .none
        }
```

cheukchuen replied:
That's right. I've included your suggestion in the latest commit.

cheukchuen committed:
Update CHANGELOG.md

cheukchuen committed:
Explicit case statements for receiver/headphones/bluetooth

jimchou-dev reviewed (approved):
Tested, LGTM
